### DISCUSSION TITLE
Translate Spicy into Python

### DISCUSSION THREAD

**Discussion by hugolin615** [zeek/spicy#1339] [Q&A] [answered]
Where should I start if I want to develop the spicy compiler to generate a parser for other programming languages, like python? I find some lex and yacc codes in spicy/toolchain/src/compiler/. Is that possible to change these codes and use gcc to create independent executables to achieve this functionality?

**Comment by bbannier** [comment, marked as answer]
> Where should I start if I want to develop the spicy compiler to generate a parser for other programming languages, like python?

Could you explain in more detail how this use case would look like?

> I find some lex and yacc codes in spicy/toolchain/src/compiler/. Is that possible to change these codes and use gcc to create independent executables to achieve this functionality?

These source files are tokenizers and parsers for Spicy source code which describe parsers written by a user.

**Reply by hugolin615** [reply]
It is mainly for some teaching materials for the course I am teaching. On top of those socket programming examples, I would like to add parsing to the application layer. When a TCP client receives an application payload, it can parse the payload. In this case, students may have a better understanding related to how parsing is working. Most students in our schools do not have a background in C and C++, so I explore the possibility of translating some protocols parsers whose structures are defined in spicy into a series of python codes that can parse them. I think that this should be done by replacing the code-generation parts of the spicy compiler.

**Reply by bbannier** [reply]
I believe what you describe is a non-trivial task. The code Spicy generates interacts and depends at runtime on libraries (runtime libraries), so even if one would add output in Python format one would still need to model these in Python as well.

I think a more manageable approach would be to wrap the generated C++ bindings for Python, see https://docs.zeek.org/projects/spicy/en/latest/host-applications.html. You might want to provide a ready solution for your students.

**Comment by hugolin615** [comment]
Thanks for the suggestions. 
Just out of curiosity,  among tokenizers and parsers at spicy/toolchain/src/compiler/ and tokenizers and parsers at hilti/toolchain/src/compiler/parser/, which ones are for the grammar of the spicy scripts?

**Reply by bbannier** [reply]
The Spicy parser is in `spicy/`, https://github.com/zeek/spicy/blob/main/spicy/toolchain/src/compiler/parser/parser.yy.
